ghcide/src/Development/IDE/GHC/Error.hs

```diff
@@ -51,30 +55,36 @@ import           Language.LSP.VFS                  (CodePointPosition (CodePoint
                                                     CodePointRange (CodePointRange))
 
 
-diagFromText :: T.Text -> D.DiagnosticSeverity -> SrcSpan -> T.Text -> FileDiagnostic
-diagFromText diagSource sev loc msg = (toNormalizedFilePath' $ fromMaybe noFilePath $ srcSpanToFilename loc,ShowDiag,)
-    Diagnostic
-    { _range    = fromMaybe noRange $ srcSpanToRange loc
-    , _severity = Just sev
-    , _source   = Just diagSource -- not shown in the IDE, but useful for ghcide developers
-    , _message  = msg
-    , _code     = Nothing
-    , _relatedInformation = Nothing
-    , _tags     = Nothing
-    , _codeDescription = Nothing
-    , _data_   = Nothing
-    }
+diagFromText :: T.Text -> D.DiagnosticSeverity -> SrcSpan -> T.Text -> Maybe (MsgEnvelope GhcMessage) -> FileDiagnostic
```

Comment:
This module is a real twisty maze of similar functions :sweat_smile: I don't have a good suggestion at the moment but it would be nice if there are fewer, since at the moment I don't feel confident that all of them are even doing the right thing. Perhaps we could rely on fewer functions even if it means making some of the call sites more verbose? I'm not sure.

Reply:
I've been thinking the same thing as I went along. The functions in `Development.IDE.Types.Diagnostics` and `Development.IDE.GHC.Error` are a mess. Lots of them restate information or pass in information twice that could be derived from a single piece of state.

That said, fixing that up feels like a relatively low priority - the current functions are a bit more complicated than they need to be, but ultimately don't take *too* long to grok IMO, and are a minimum of changes necessary to get the system working without changing many call sites too much.

Could we push to a ticket, another PR? Considering I'm the most likely candidate for doing this, I'm alright with that.

Reply:
(In the meantime, reverted)

Reply:
Looking deeper, `attachReason` was introduced in f03a7fa55d909c864771f0989abace8793b5b15a, explicitly for code actions that would disable a warning.

If we reimplement the code action in some other way, then we should be good.

Reply:
Fine with me. It's just hard for me to review since I'm failing to grok them D: